Add macro to check error code additions/combinations

chris-jones-arm opened the pull request:
## Description
This PR adds a new macro `MBEDTLS_ERR_ADD` to aid in testing that error codes are correctly combined together. This macro performs a check, when `MBEDTLS_TEST_HOOKS` is enabled, to determine whether the returned error code would be corrupted. If the returned error code will be corrupted the test will fail and report the location of the error otherwise there is no change in behaviour.

`MBEDTLS_ERR_ADD` will replace all occurrences of error code addition/combination in the codebase so that these errors (such as the one found in #3911) can be found before being introduced.

## Status
**IN DEVELOPMENT**

## Requires Backporting
Unsure.

## Additional comments
While this does not fix the example given in #3911, it does give us greater confidence that incorrect error codes will not be added in the future.

## Todos
- [ ] Add changelog
- [ ] Replace other occurrences
- [ ] Backports?


## Steps to test or reproduce
- Enable `MBEDTLS_TEST_HOOKS` (`./scripts/config.pl set MBEDTLS_TEST_HOOKS`).
- Run tests

chris-jones-arm commented:
I am almost certain this will need a changelog entry however I am unsure which catagory this change falls under, any advice?

gilles-peskine-arm commented on library/common.c:
```diff
@@ -0,0 +1,27 @@
+/*
+ *  Internal invasive testing helper functions
```
I'm not convinced that a `common.c` is the right place for `mbedtls_err_add`.

Given that this is a test hook, we could live with having it in `error.c` and only enabled when error messages are enabled.

In any case, `common.c` is not an acceptable name for the sake of Mbed OS (Mbed TLS is periodically imported into the Mbed OS source tree). The Mbed OS build system has this peculiarity that it doesn't accept two `.c` files with the same name, even in different directories belonging to different subprojects. So we shouldn't use names that are too generic for `library/*.c`. Anything that's obviously crypto/TLS related is fine (e.g. `aes.c`, `ssl_*.c`, `x509_*.c`). We got away with `debug.c` but it's not something we would have created after Mbed OS became relevant. Let's not add `common.c` even if nobody has nabbed it yet (I haven't checked).

chris-jones-arm replied:
I [originally](https://github.com/ARMmbed/mbedtls/pull/4006/commits/f6b76e3ba8326dd40c4870dc884a6fdf087356a9) had placed `mbedtls_err_add` in `error.c` however I moved it when I realised that `error.c` was an auto-generated file. I will look into adding it through the script that generates it.

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on library/common.h:
```diff
@@ -44,10 +45,25 @@
-#endif
+
+#define MBEDTLS_ERR_ADD( high, low ) \
+    ( high + low )
```
Macro hygiene:
```suggestion
    ( ( high ) + ( low ) )
```

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on library/common.h:
```diff
@@ -29,6 +29,7 @@
 #include "mbedtls/config.h"
 #endif
 
+#if defined(MBEDTLS_TEST_HOOKS)
```
The file would be more readable if it was structured by declaration than by compilation option. Also, the documentation here applies whether or not `MBEDTLS_TEST_HOOKS` is enabled. So please keep the structure the way it was before.

```
/** documentation of MBEDTLS_STATIC_TESTABLE */
#if defined(MBEDTLS_TEST_HOOKS)
#define MBEDTLS_STATIC_TESTABLE
#else
#define MBEDTLS_STATIC_TESTABLE static
#endif

/** documentation of MBEDTLS_ERR_ADD */
#if defined(MBEDTLS_TEST_HOOKS)
#define MBEDTLS_ERR_ADD mbedtls_err_add()
void mbedtls_err_add();
#else
#define MBEDTLS_ERR_ADD (high+low)
#endif
```

chris-jones-arm replied:
Thanks, I wasn't sure of myself when I made the change originally. I will make the change now.

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on tests/src/helpers.c:
```diff
@@ -244,3 +244,16 @@ void mbedtls_param_failed( const char *failure_condition,
+{
+    if ( high > -0x1000 || low < -0x007F )
+    {
+        mbedtls_fprintf( stderr, "\nIncorrect error code addition at %s:%d\n",
+                                file, line );
+        mbedtls_exit( 1 );
```
This is very unfriendly. The test case won't be reported as failed, and subsequent test cases won't run.

The right thing to do is to mark the test case as failed. At least, I think that's a good approximation: the error location will be at the place where `MBEDTLS_ERR_ADD` is called rather than in the test suite, so the programmer might end up needing a debugger to figure out how the problematic function was called, but we can't easily do better.

Marking the test as failed is annoyingly difficult because `test_fail` is defined in `helpers.function` and therefore can't be called from `tests/src/helpers.c`. I guess it's time to move `test_fail` and friends to a new file in `tests/src`, or maybe `tests/src/helpers.c` itself. This is a natural follow-up to https://github.com/ARMmbed/mbedtls/pull/3387 that we just haven't got around to yet. This is not the first time we've hit this limitation with `test_fail`, for example it bit me recently in https://github.com/ARMmbed/mbedtls/pull/3912#discussion_r555668044.

Chris, do you have the time to do the move of `test_fail` and friends (maybe chat with Ronald or me first to agree on a scope), then rebase `mbedtls_test_err_add_check` on top of that?

chris-jones-arm replied:
Yes I agree this is not how I wanted to do it and this was simply a temporary workaround of that issue. I was going to use `TEST_HELPER_ASSERT` as that is in scope of `helpers.c` however decided against it as it also doesn't report the correct location of the failure as well as just exiting instead of failing the test.

I will take a look at moving `test_fail` now as I'd be happy to do the work.

chris-jones-arm replied:
#4028 and #4054 have been created to address this issue.

chris-jones-arm replied:
#4054 has been merged and this PR has been rebased on top to now include these changes.

I have also added a commit to use `mbedtls_test_fail` as opposed to manually printing and exiting as `mbedtls_test_fail` is now available.

gilles-peskine-arm commented on tests/src/helpers.c:
```diff
@@ -244,3 +244,16 @@ void mbedtls_param_failed( const char *failure_condition,
+void mbedtls_test_err_add_check( int high, int low,
+                                 const char *file, int line )
+{
+    if ( high > -0x1000 || low < -0x007F )
```
These checks are incomplete.

* `high` and `low` should be negative or zero. And I'm not even sure about allowing zero.
* `high` should not contain a low-level error code already.

chris-jones-arm replied:
I'm not sure I completely agree with that statement.

- `high` is always negative and therefore cannot be zero. `low` on the other hand can be positive and hence can be zero, I am happy to add this check. I completely agree that zero should not be allowed as it would not make sense to add a success to an error (that is still an error).

- "high should not contain a low-level error code already."
I'm not entirely sure what you mean by this. The check throws an error if `high` is negatively smaller than a high error code and thus any low error code will be found to be incorrect and therefore the check will throw an error.

gilles-peskine-arm replied:
`high=-0x2104` should be not allowed. That's not a high-level error code, it's already a combination of a high-level error code with a low-level error code.

chris-jones-arm replied:
Ah okay, thanks for the clarification. I will make the changes now.

chris-jones-arm replied:
Just so you are aware in my next push I have allowed zero to pass the check as I found a case (`rsa.c:1282`) where `ret` can either be zero or an error. This could be something to change in the future but I don't think it makes enough of a difference for us to care about now.

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on tests/src/helpers.c:
```diff
@@ -244,3 +244,16 @@ void mbedtls_param_failed( const char *failure_condition,
+{
+    if ( high > -0x1000 || low < -0x007F )
+    {
+        mbedtls_fprintf( stderr, "\nIncorrect error code addition at %s:%d\n",
```
It would be nicer to have different error messages for each condition (positive high, positive low, high contains low bits, low contains high bits, high is outside 16-bit range, low is outside 16-bit range).

chris-jones-arm replied:
Done.

gilles-peskine-arm commented:
I don't see any need for a changelog. This doesn't affect users of the library. We don't write changelog entries for testing or robustness improvements.

ronald-cron-arm commented on library/common.c:
```diff
@@ -0,0 +1,27 @@
+ */
+
+#if defined(MBEDTLS_TEST_HOOKS)
+void (*mbedtls_test_err_add_hook)( int, int, const char *, int );
```
This should be in a test header and the test header included in this file.

mpg commented:
It looks to me like there are 2 reviewers already, so I'm removing the label "needs: reviewer". Please add it back if I was mistaken.

gilles-peskine-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,32 @@ void mbedtls_param_failed( const char *failure_condition,
+void mbedtls_test_err_add_check( int high, int low,
+                                 const char *file, int line )
+{
+    if ( high > -0x1000 )
```
I think that apart from `low > 0`, the checks are complete and correct. However:

* They would be easier to read with comparisons of positive numbers, and for some of them, with bit masks. A {high,low}-level error code is a value whose opposite has only the bits {7–14,0–6} set. We don't currrently use high-level error codes whose upper 3 bits are 0, but `error.h` still defines them as high-level error codes.

chris-jones-arm replied:
Done.

mpg commented on include/mbedtls/error.h:
```diff
@@ -114,6 +114,26 @@ extern "C" {
+ *
+ * \param hook    hook to invasive testing function
+ */
+void mbedtls_set_err_add_hook( void (*hook)( int, int, const char *, int ) );
+int mbedtls_err_add( int high, int low, const char *file, int line );
+#define MBEDTLS_ERR_ADD( high, low )  \
+    ( mbedtls_err_add( high, low, __FILE__, __LINE__ ) )
+#else
+#define MBEDTLS_ERR_ADD( high, low ) \
+    ( ( high ) + ( low ) )
```
We do use `static inline` in 2.16 and 2.7 (git grep finds 59 and 49 occurrences), that's one of the few C99 things that were already allowed (together with `stdint.h`). We just have a compatibility define for armcc5 and MSVC:
```
#if ( defined(__ARMCC_VERSION) || defined(_MSC_VER) ) && \
    !defined(inline) && !defined(__cplusplus)
#define inline __inline
#endif
```
(which by the way is also present in development).

chris-jones-arm replied:
In my most recent commits I have switched over to using this version of static inline function. Unfortunately it seems that the only way to keep the file and line number debug capabilities is to use a wrapper macro (`mbedtls_error_add`) which adds on the relevant info. Hopefully this solution is acceptable.

I will also mention this macro meant changing the actual function name to `mbedtls_error_add_ext` which I don't particularly like however I could not think of anything better at the time.

chris-jones-arm commented on tests/include/test/helpers.h:
```diff
@@ -260,4 +260,14 @@ void mbedtls_test_param_failed_reset_state( void );
 
+#if defined(MBEDTLS_TEST_HOOKS)
+/**
+ * \brief   Check that a pure high-level error code is being combined with a
```
Thinking about this some more, I am sure that neither high or low level error codes can be positive (greater than zero) however the case of zero is somewhat harder.

Currently low level error codes are allowed to be zero as there are combinations (in `rsa.c` and possibly elsewhere) where a high level error needs to be returned as well as a `ret` which *may or may not* contain a low level error code. I think this makes sense and should be allowed as it significantly reduces the effort to combine error codes in the codebase.

I am not so sure on whether high level error codes should be allowed to be equal to zero however. I cannot see anywhere that it provides benefit as most often a low level function (that can return a low level error) is called by a high level one that will always want to return its own error or will just return the low level error without wanting to do any combination. As this is not specified explicitly by `error.h` though I am thinking it may be easier to make one extra case where `high` is allowed to be zero.

@gilles-peskine-arm what are your thoughts on this?

chris-jones-arm replied:
Actually don't worry, currently it is possible to combine zero with a low level error code by doing `0 + MBEDTLS_ERR_XXX` and there is no reason why when combining error codes we cannot do the same even if it isn't currently done in the library. I will add another check for `high == 0` now.

chris-jones-arm replied:
Done.

chris-jones-arm commented on tests/include/test/helpers.h:
```diff
@@ -260,4 +260,14 @@ void mbedtls_test_param_failed_reset_state( void );
 
+#if defined(MBEDTLS_TEST_HOOKS)
+/**
+ * \brief   Check that a pure high-level error code is being combined with a
```
Done.

chris-jones-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,32 @@ void mbedtls_param_failed( const char *failure_condition,
+    {
+        mbedtls_test_fail( "'low' is greater than 8-bits", line, file );
+    }
+    else if ( low > 0 )
```
Done.

chris-jones-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,32 @@ void mbedtls_param_failed( const char *failure_condition,
+    }
+    else if ( low < -0x007F )
+    {
+        mbedtls_test_fail( "'low' is greater than 8-bits", line, file );
```
Done.

chris-jones-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,32 @@ void mbedtls_param_failed( const char *failure_condition,
+    }
+    else if ( high < -0x7F80 )
+    {
+        mbedtls_test_fail( "'high' is greater than 16-bits", line, file );
```
Done.

chris-jones-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,32 @@ void mbedtls_param_failed( const char *failure_condition,
+    }
+    else if ( high < -0x7F80 )
+    {
+        mbedtls_test_fail( "'high' is greater than 16-bits", line, file );
```
Done.

gilles-peskine-arm commented on tests/src/helpers.c:
```diff
@@ -287,27 +287,41 @@ void mbedtls_param_failed( const char *failure_condition,
+                            line, file );
     }
-    else if ( ( high & 0x7F ) != 0 )
+    else if ( ( high & 0x7F ) != 0 )    // high & 0000000011111111
```
Non-blocker: bitwise operations on negative numbers aren't always portable. In this particular case, it works except on one's complement machines (the behavior of `&` is fully specified, but the representations of the operands depend on the negative integer representation, and the operation only does what we want with two's complement and sign-magnitude). I don't care about a hypothetical machine that uses one's complement and is otherwise capable of running Mbed TLS: it's unlikely that there are or ever will be any. But avoiding bitwise operations on negative numbers would make the code easier to read: I'd prefer not to have to think about these things when I'm reading the code.

gilles-peskine-arm commented on include/mbedtls/error.h:
```diff
@@ -114,25 +114,47 @@ extern "C" {
+ * \brief Testing hook called before adding/combining two error codes together.
+ *        Only used when invasive testing is enabled via MBEDTLS_TEST_HOOKS.
+ */
+extern void (*mbedtls_test_hook_error_add)( int, int, const char *, int );
```
This definition and the corresponding declaration should be wrapped in `#if defined(MBEDTLS_TEST_HOOKS)`.

chris-jones-arm replied:
I agree that this definition should be wrapped however the following declaration (I assume you mean `mbedtls_err_add_ext`) should not be because it needs to be visible to `MBEDTLS_ERROR_ADD` even when `MBEDTLS_TEST_HOOKS` is not defined.

This was precisely to move away from having two different definitions of `MBEDTLS_ERROR_ADD` as we had before, see: https://github.com/ARMmbed/mbedtls/pull/4006#discussion_r576352099.

gilles-peskine-arm replied:
Sorry, I meant “this declaration and the corresponding definition” (or I meant to post this comment on the definition of `mbedtls_test_hook_error_add` in `error.c`. Either way d33be30564b37868dbe7c18a93c590d1da1011a0 does what I meant.

gilles-peskine-arm commented on include/mbedtls/error.h:
```diff
@@ -114,25 +114,47 @@ extern "C" {
+ *        called via the mbedtls_error_add macro.
  *
- * \param hook    hook to invasive testing function
+ * \note  When invasive testing is enabled via MBEDTLS_TEST_HOOKS also try to
```
```suggestion
 * \note  When invasive testing is enabled via #MBEDTLS_TEST_HOOKS, also try to
```

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on include/mbedtls/error.h:
```diff
@@ -114,25 +114,47 @@ extern "C" {
  *
- * \param hook    hook to invasive testing function
+ * \note  When invasive testing is enabled via MBEDTLS_TEST_HOOKS also try to
+ *        call mbedtls_test_hook_error_add.
```
```suggestion
 *        call \ink mbedtls_test_hook_error_add.
```
(according to https://stackoverflow.com/questions/17395098/doxygen-how-to-link-to-a-c-file-scope-global-variable, please check if it works — if you can't make it work quickly, just leave it unlinked and use `\c`)

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on include/mbedtls/error.h:
```diff
@@ -114,25 +114,47 @@ extern "C" {
- * This hook is used in the test infrastructure to report on errors when
- * combining two error codes of the same level.
+ *        This function can be called directly however it is usually
+ *        called via the mbedtls_error_add macro.
```
```suggestion
 *        called via the #MBEDTLS_ERROR_ADD macro.
```

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on include/mbedtls/error.h:
```diff
@@ -114,6 +114,48 @@ extern "C" {
 #define MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED -0x006E  /**< This is a bug in the library */
 
+/**
+ * \brief Combines a high-level and low-level error code together.
```
Please document that it's ok for `high`, `low` or both to be 0.

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,46 @@ void mbedtls_param_failed( const char *failure_condition,
+     * examples given in comments have been made positive for ease of
+     * understanding. The structure of an error code is such:
+     *
+     *                                                shhhhhhhllllllll
```
```suggestion
     *                                                shhhhhhhhlllllll
```
and some of the base 2 values below are similarly off by one

chris-jones-arm replied:
Done.

gilles-peskine-arm reviewed (changes requested):
Looks mostly good to me. The remaining issues are mostly in the documentation.

gilles-peskine-arm commented:
Unfortunately there's a merge conflict. Please rebase or merge.

chris-jones-arm commented:
I have now rebased onto development and added two extra commits. One applies the use of the macro to the rest of the library and the other fixes a bug already caught by the new invasive tests on error code addition.

gilles-peskine-arm commented:
Builds are failing

chris-jones-arm commented:
CI error is transient and can be considered passing.

gilles-peskine-arm reviewed (approved):
Looks good to me.

I have not reviewed the change of `+` to `MBEDTLS_ERROR_ADD` for completeness, only for correctness. I don't think it's useful to insist for this change to be complete because there are likely pull requests in flight that make such additions. We should wait a bit for such pull requests to be merged or get stale, then do a new round of hunting for error code additions and perhaps make that enforced by a CI check.

ronald-cron-arm commented on include/mbedtls/error.h:
```diff
@@ -114,6 +119,54 @@ extern "C" {
+ * \param file      file where this error code addition occured.
+ * \param line      line where this error code addition occured.
+ */
+static inline int mbedtls_error_add_ext( int high, int low,
```
Out of curiosity why the "_ext" suffix?

chris-jones-arm replied:
See https://github.com/ARMmbed/mbedtls/pull/4006#discussion_r605738001.

Basically I could not think of a better name and it had to be different to `mbedtls_error_add`. If you have a better name please let me know and I'd be happy to change it.

chris-jones-arm replied:
Actually now that the macro is capitalised (`MBEDTLS_ERROR_ADD`) I suppose it could be `mbedtls_error_add`. @ronald-cron-arm Would you like me to make this change?

ronald-cron-arm replied:
Yes please.

ronald-cron-arm reviewed (changes requested):
I have a few minor comments.

chris-jones-arm commented on include/mbedtls/error.h:
```diff
@@ -114,6 +119,54 @@ extern "C" {
+ * \brief Testing hook called before adding/combining two error codes together.
+ *        Only used when invasive testing is enabled via MBEDTLS_TEST_HOOKS.
+ */
+extern void (*mbedtls_test_hook_error_add)( int, int, const char *, int );
```
I think it was done because the testing framework has visibility/access to the library but not the other way around. This way the testing module sets it's own checking function to be called when an addition occurs and `error.h` does not need to include test specific code.

It also means that it will still run (without checks) for users who do not run our testing framework unless they wanted to use their own checking function.

Whether this is actually useful and something we want is beyond me so if you still disagree @gilles-peskine-arm may have a better idea of why it's important.

chris-jones-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,46 @@ void mbedtls_param_failed( const char *failure_condition,
+     * h = high level error code (includes high and module error codes).
+     * l = low level error code.
+     */
+    if ( high > -0x1000 && high != 0 )  // high < 0001000000000000
```
I admit that it is hard to read hence the existing comment, I will try to clarify this better now. FYI it is checking that the variable `high` contains some high-level error code bits ("High-level module nr (3 bits - 0x0...-0x7...)" from `error.h`). If `high` is less than the first high level error code bit, and is not zero, then it is not a high level error code and fails.

chris-jones-arm replied:
Done.

chris-jones-arm commented on include/mbedtls/error.h:
```diff
@@ -114,6 +119,54 @@ extern "C" {
+/**
+ * \brief Combines a high-level and low-level error code together.
+ *
+ *        Wrapper function for mbedtls_err_add_ext(). See that function for
```
Done.

chris-jones-arm commented on library/ssl_cookie.c:
```diff
@@ -220,8 +220,8 @@ int mbedtls_ssl_cookie_check( void *p_ctx,
     if( mbedtls_mutex_unlock( &ctx->mutex ) != 0 )
-        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR +
-                MBEDTLS_ERR_THREADING_MUTEX_ERROR );
+        return( MBEDTLS_ERROR_ADD( MBEDTLS_ERR_SSL_INTERNAL_ERROR ,
```
Done.

ronald-cron-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,58 @@ void mbedtls_param_failed( const char *failure_condition,
+     */
+    if ( high > -0x1000 && high != 0 )
+    /* high < 0001000000000000
+     * No high level error bits are set.
```
```suggestion
     * No high level module ID bits are set.
```

chris-jones-arm replied:
No I do mean "high level error bits" here. We cannot have an error code with just module ID bits without any high level error bits set.

ronald-cron-arm replied:
I am referring to the documentation in error.h for the "naming of the bits":
```c 
16 bit error code bit-segmentation                                           
                                                                               
1 bit  - Unused (sign bit)                                                   
3 bits - High level module ID                                                
5 bits - Module-dependent error code                                         
7 bits - Low level module errors
```
I think we are talking about the same thing here a.k.a bits 12..14. It is just that according to the doc copied above the proper name for those bits is "High level module ID".

chris-jones-arm replied:
Thanks for the clarification, I missed their precise name in the documentation. Now fixed.

ronald-cron-arm commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,58 @@ void mbedtls_param_failed( const char *failure_condition,
+     *                                                shhhhhhhhlllllll
+     *
+     * s = sign bit.
+     * h = high level error code (includes high and module error codes).
```
```suggestion
     * h = high level error code (includes high level module ID (bits 12..14) and module-dependent error code (bits 7..11).
```

chris-jones-arm replied:
Done.

chris-jones-arm replied:
Done.

gilles-peskine-arm commented on tests/src/helpers.c:
```diff
@@ -295,25 +295,38 @@ void mbedtls_test_err_add_check( int high, int low,
      *
      * s = sign bit.
-     * h = high level error code (includes high and module error codes).
+     * h = high level error code (includes high level module ID (bits 12..14)
+     *     and module-dependent error code (bits 7..11)).
```
Bits 7..11 and 12..14 aren't really separate anymore: we ran out of “module IDs” a while ago, and SSL overflowed. Not a blocker since you're just reflecting the somewhat obsolete description in `error.h`.

chris-jones-arm replied:
I agree that they aren't really separate however as we have not updated the documentation in `error.h` I consider this out of scope for this PR.

gilles-peskine-arm commented on tests/src/helpers.c:
```diff
@@ -295,25 +295,38 @@ void mbedtls_test_err_add_check( int high, int low,
-    else if ( high < -0x7F80 )          // high > 0111111110000000
+    else if ( high < -0x7F80 )
+    /* high > 0111111110000000
+     * Error code is larger than the greatest high + module level error.
```
I don't know what “greatest high + module level error” means (I don't even know how to parse it). I don't actually need an explanation here, I follow what the code is doing, but if the only way to understand the comment is to read the code, the comment isn't right.

chris-jones-arm replied:
I think I've clarified this a bit better, hopefully it is easier to understand now.

gilles-peskine-arm reviewed (changes requested):
Ok except that I find some of the new comments more confusing than helpful.

ronald-cron-arm commented:
@chris-jones-arm @gilles-peskine-arm Do we need to backport this PR to development_2.x?

gilles-peskine-arm commented:
Yes, this is supposed to go into 2.x. It's based on 2.x so the same branch can go into both, but we still need a separate PR because we do all our merging via GitHub. I'll make one.

gilles-peskine-arm commented:
We do need at least a partial 2.16 backport for the bug fix. Do we backport the rest?

* Replacing `+` by `MBEDTLS_ERROR_ADD`: would make backports easier. Low risk of behavior change. Should have no impact on code size.
* Defining `MBEDTLS_ERROR_ADD` as something other than `+`: maybe? `MBEDTLS_TEST_HOOKS` is in 2.16. But the test code structure is very different so backporting the test suite part might be too much work to bother.
*

chris-jones-arm commented:
While I think it's a good idea to keep consistency across branches I suspect in this case the changes would take too much effort and I'd like to not delay this being merged. In any case as we are not introducing new features to 2.16 I suspect there would be few (if not non-existent) errors that this catches on 2.16 that the original tests on `development` would not catch.

mpg commented:
I only have a clear opinion on two points:

1. Clearly the bug fixes have to be backported
2. I don't think backporting the improvements to the test framework is worth it, as Chris says this is an LTS.

I'm more on the fence regarding "Replacing `+` by `MBEDTLS_ERROR_ADD`". As Gilles said, this is likely to make future backports easier. 2.16 has about 1 year of support left. As a point comparison, 2.7 had 48 ChangeLog entries (so I assume roughly the same number of backport PRs) in its last year of support. Of course, if we assume 2.16 is going to get about 50 backports before it's EOL, it's hard to predict how many would have conflicts that could be avoided by backporting this change, but my guess would be that backporting that part will be worth it in the long run.

gilles-peskine-arm commented:
>  2.16 has about 1 year of support left. As a point comparison, 2.7 had 48 ChangeLog entries (so I assume roughly the same number of backport PRs) in its last year of support. Of course, if we assume 2.16 is going to get about 50 backports before it's EOL, it's hard to predict how many would have conflicts that could be avoided by backporting this change, but my guess would be that backporting that part will be worth it in the long run.

Oh, looking at historical data is a great idea!

```
git checkout mbedtls-2.16
git log --format=%H --since='1 year ago' -G 'MBEDTLS_ERR_\w+ \+|\+ MBEDTLS_ERR_'
```
This only returns 3 commits. This might miss places with different formatting, and it definitely misses places where there was a change that was adjacent to a line with this content, but not on the line with the addition itself. But even so, that's not much.

And if for whatever reason we end up backporting a patch that affects a lot of error code additions, we can backport the definition of `MBEDTLS_ERROR_ADD` later.

So let's just backport the bug fix.

mpg commented:
I think the following git command includes the commits when only context has changed:
```
git log --patch --since='1 year ago' | egrep 'MBEDTLS_ERR_\w+ \+|\+ MBEDTLS_ERR_'

gilles-peskine-arm commented:
That's settled then, I'll make the one-commit 2.16 backport.

gilles-peskine-arm commented on library/pkparse.c:
```diff
@@ -1028,26 +1028,35 @@ static int pk_parse_key_pkcs8_unencrypted_der(
     if( ( ret = pk_get_pk_alg( &p, end, &pk_alg, &params ) ) != 0 )
-        return( MBEDTLS_ERR_PK_KEY_INVALID_FORMAT + ret );
+    {
+        if( ret >= -0x007F )
```
This actually can't happen. `pk_get_pk_alg` always returns a PK error code. So we should just return `ret`.

chris-jones-arm replied:
Good spot! I have confirmed and fixed now.

gilles-peskine-arm reviewed (changes requested):
I just noticed that the bugfix in `pkparse` adds some dead code.
